File: src/server/discovery/udp.py
"""UDP discovery server.

Runs one background thread that broadcasts "MAG_SERVER <ip> <port>" to
255.255.255.255:<udp_port> every 5 seconds.  Clients listen passively for
this packet; they send nothing.  This keeps LAN traffic constant at one
packet per 5 seconds regardless of how many students are connected.
"""

import socket
import threading

from server.logging import get_logger

_LOG = get_logger(__name__)
_BROADCAST_INTERVAL = 5.0


def _local_ip() -> str:
    """Best-effort: connect a UDP socket to a public address to find the
    local IP that the OS would use for LAN traffic."""
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("8.8.8.8", 80))
        return s.getsockname()[0]
    except OSError:
        return "127.0.0.1"
    finally:
        s.close()


class UDPDiscovery:
    def __init__(self, udp_port: int, http_port: int) -> None:
        self._udp_port = udp_port
        self._http_port = http_port
        self._ip = _local_ip()
        self._announcement = f"MAG_SERVER {self._ip} {self._http_port}".encode()
        self._stop = threading.Event()
        self._broadcaster = threading.Thread(target=self._broadcast_loop, daemon=True)

    @property
    def server_ip(self) -> str:
        return self._ip

    def start(self) -> None:
        self._broadcaster.start()
        _LOG.info("UDP discovery started on %s:%d", self._ip, self._udp_port)

    def stop(self) -> None:
        self._stop.set()

    def _broadcast_loop(self) -> None:
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        try:
            while not self._stop.is_set():
                try:
                    sock.sendto(self._announcement, ("255.255.255.255", self._udp_port))
                    _LOG.debug("Broadcast: %s", self._announcement.decode())
                except OSError as e:
                    _LOG.warning("Broadcast error: %s", e)
                self._stop.wait(_BROADCAST_INTERVAL)
        finally:
            sock.close()

File: src/client/net/udp.cpp
#include "network.h"

#include <string>
#include <optional>
#include <chrono>
#include <thread>
#include <sstream>

#ifdef _WIN32
#  include <winsock2.h>
#  include <ws2tcpip.h>
#  pragma comment(lib, "ws2_32.lib")
using sock_t = SOCKET;
constexpr sock_t INVALID = INVALID_SOCKET;
static void sock_close(sock_t s) { closesocket(s); }
#else
#  include <sys/socket.h>
#  include <netinet/in.h>
#  include <arpa/inet.h>
#  include <unistd.h>
#  include <fcntl.h>
using sock_t = int;
constexpr sock_t INVALID = -1;
static void sock_close(sock_t s) { close(s); }
#endif

namespace mag {

namespace {

// Parse "MAG_SERVER <ip> <port>" into a ServerInfo.
std::optional<ServerInfo> parse_server_msg(const char* buf, int len) {
    std::string msg(buf, len);
    if (msg.rfind("MAG_SERVER ", 0) != 0) return std::nullopt;
    std::istringstream ss(msg.substr(11));
    ServerInfo info;
    int port;
    if (ss >> info.ip >> port) {
        info.port = port;
        return info;
    }
    return std::nullopt;
}

#ifdef _WIN32
struct WsaGuard {
    WsaGuard() { WSADATA d; WSAStartup(MAKEWORD(2, 2), &d); }
    ~WsaGuard() { WSACleanup(); }
};
static WsaGuard _wsa_guard;
#endif

} // namespace

std::optional<ServerInfo> discover_server(int udp_port, int timeout_sec) {
    sock_t sock = socket(AF_INET, SOCK_DGRAM, 0);
    if (sock == INVALID) return std::nullopt;

    int yes = 1;
    setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, reinterpret_cast<const char*>(&yes), sizeof(yes));

    // Bind to receive the server's periodic broadcast.
    sockaddr_in local{};
    local.sin_family = AF_INET;
    local.sin_port = htons(static_cast<uint16_t>(udp_port));
    local.sin_addr.s_addr = INADDR_ANY;
    if (bind(sock, reinterpret_cast<sockaddr*>(&local), sizeof(local)) < 0) {
        sock_close(sock);
        return std::nullopt;
    }

    // 1 s recv timeout so the timeout_sec check runs at least once per second.
#ifdef _WIN32
    DWORD tv_ms = 1000;
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, reinterpret_cast<const char*>(&tv_ms), sizeof(tv_ms));
#else
    timeval tv{1, 0};
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
#endif

    // Passively listen for the server's periodic broadcast.
    // The server sends "MAG_SERVER <ip> <port>" every 5 s; clients send nothing,
    // keeping LAN traffic constant regardless of how many students are present.
    auto start = std::chrono::steady_clock::now();

    while (true) {
        char buf[256]{};
        sockaddr_in from{};
        socklen_t fromlen = sizeof(from);
        int n = recvfrom(sock, buf, sizeof(buf) - 1, 0,
                         reinterpret_cast<sockaddr*>(&from), &fromlen);
        if (n > 0) {
            auto info = parse_server_msg(buf, n);
            if (info) {
                sock_close(sock);
                return info;
            }
        }
        if (timeout_sec > 0) {
            auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(
                std::chrono::steady_clock::now() - start).count();
            if (elapsed >= timeout_sec) {
                sock_close(sock);
                return std::nullopt;
            }
        }
    }
}

} // namespace mag
